src/components/registration/steps.jsx

```diff
@@ -82,38 +82,63 @@ module.exports = {
         onChangeShowPassword: function (field, value) {
             this.setState({showPassword: value});
         },
-        onValidSubmit: function (formData, reset, invalidate) {
-            this.setState({waiting: true});
+        validateUsername: function (username, callback) {
             api({
                 host: '',
-                uri: '/accounts/check_username/' + formData.user.username + '/'
-            }, function (err, res) {
+                uri: '/accounts/check_username/' + username + '/'
+            }, function (err, body, res) {
                 var formatMessage = this.props.intl.formatMessage;
-                this.setState({waiting: false});
-                if (err) return invalidate({all: err});
-                res = res[0];
-                switch (res.msg) {
+                if (err || res.statusCode !== 200) {
+                    err = err || formatMessage({id: 'general.error'});
+                    this.refs.form.refs.formsy.updateInputsWithError({all: err});
+                    if (typeof callback !== 'undefined') {
+                        callback(false);
+                    }
+                    return;
+                }
+                body = body[0];
+                var isValid = false;
+
+                switch (body.msg) {
                 case 'valid username':
                     this.setState({
                         validUsername: 'pass'
                     });
-                    return this.props.onNextStep(formData);
+                    isValid = true;
+                    break;
                 case 'username exists':
-                    return invalidate({
+                    this.refs.form.refs.formsy.updateInputsWithError({
                         'user.username': formatMessage({id: 'registration.validationUsernameExists'})
                     });
+                    break;
                 case 'bad username':
-                    return invalidate({
+                    this.refs.form.refs.formsy.updateInputsWithError({
                         'user.username': formatMessage({id: 'registration.validationUsernameVulgar'})
                     });
+                    break;
                 case 'invalid username':
                 default:
-                    return invalidate({
+                    this.refs.form.refs.formsy.updateInputsWithError({
                         'user.username': formatMessage({id: 'registration.validationUsernameInvalid'})
                     });
+                    break;
+                }
+
+                if (typeof callback !== 'undefined') {
+                    callback(isValid);
```

Comment:
One last suggestion that I think is just stylistic — if you make `callback` optional on the first line of the function with `callback = callback || function () {};`, then you can always just `return callback(true/false)` without checking the type of `callback`.

If the reason you changed the style here was to avoid repeating the `if (typeof callback !== 'undefined') {...}` in every `case`, then this would allow you to just return within each `case` too. Personally I find the inline returns more readable than the fall-through style, since you know what happens as soon as you see a `return`.